File: packages/lessmark/src/markdown.js
import { formatAst } from "./format.js";
import { parseLessmark } from "./parser.js";
import { DECISION_ID_PATTERN, isSafeHref, isSafeResource, splitTableRow } from "./rules.js";

const MAX_INLINE_MARKDOWN_PASSES = 128;

export function fromMarkdown(markdown) {
  const lines = String(markdown).replace(/^\uFEFF/, "").replace(/\r\n?/g, "\n").split("\n");
  const children = [];
  let index = 0;

  while (index < lines.length) {
    const line = lines[index];

    if (line.trim() === "") {
      index += 1;
      continue;
    }

    const heading = /^(#{1,6})\s+(.+?)\s*$/.exec(line);
    if (heading) {
      children.push({ type: "heading", level: heading[1].length, text: plainText(heading[2]) });
      index += 1;
      continue;
    }

    if (isMarkdownSeparator(line)) {
      children.push({ type: "block", name: "separator", attrs: {}, text: "" });
      index += 1;
      continue;
    }

    if (readMathFenceLine(line)) {
      const body = [];
      index += 1;
      while (index < lines.length && !readMathFenceLine(lines[index])) {
        body.push(lines[index]);
        index += 1;
      }
      if (index >= lines.length) {
        throw new Error("Unclosed math block");
      }
      index += 1;
      children.push({ type: "block", name: "math", attrs: { notation: "tex" }, text: body.join("\n") });
      continue;
    }

    const fence = readFenceLine(line);
    if (fence) {
      const body = [];
      index += 1;
      while (index < lines.length && !isClosingFence(lines[index], fence)) {
        body.push(lines[index]);
        index += 1;
      }
      if (index >= lines.length) {
        throw new Error("Unclosed fenced code block");
      }
      index += 1;
      const fenced = fencedCodeNode(fence.lang, body);
      children.push(fenced);
      continue;
    }

    const task = /^\s*[-*]\s+\[([ xX])\]\s+(.+?)\s*$/.exec(line);
    if (task) {
      children.push({
        type: "block",
        name: "task",
        attrs: { status: task[1].toLowerCase() === "x" ? "done" : "todo" },
        text: plainText(task[2])
      });
      index += 1;
      continue;
    }

    const list = readMarkdownList(lines, index);
    if (list) {
      children.push(list.node);
      index = list.nextIndex;
      continue;
    }

    const image = readImageLine(line);
    if (image) {
      if (isSafeResource(image.src)) {
        const attrs = { src: image.src, alt: plainText(image.alt) || "Image" };
        if (image.caption) attrs.caption = plainText(image.caption);
        children.push({ type: "block", name: "image", attrs, text: "" });
      } else {
        children.push({ type: "block", name: "paragraph", attrs: {}, text: plainText(image.alt) });
      }
      index += 1;
      continue;
    }

    const quote = readBlockquote(lines, index);
    if (quote) {
      children.push(quote.node);
      index = quote.nextIndex;
      continue;
    }

    const table = readTable(lines, index);
    if (table) {
      children.push(table.node);
      index = table.nextIndex;
      continue;
    }

    const paragraph = [];
    while (index < lines.length && lines[index].trim() !== "") {
      if (isMarkdownBlockStart(lines, index)) {
        break;
      }
      paragraph.push(lines[index].trim());
      index += 1;
    }

    const rawText = paragraph.join(" ");
    const text = plainText(rawText);
    const link = /^\[([^\]]+)\]\(([^)\s]+)\)$/.exec(rawText);
    if (link && isSafeHref(link[2])) {
      children.push({ type: "block", name: "link", attrs: { href: link[2] }, text: link[1] });
    } else {
      children.push({ type: "block", name: "paragraph", attrs: {}, text });
    }
  }

  return formatAst({ type: "document", children });
}

export function toMarkdown(lessmark) {
  const ast = typeof lessmark === "string" ? parseLessmark(lessmark) : lessmark;
  const footnoteIds = collectFootnoteIds(ast);
  const chunks = ast.children.map((node) => {
    if (node.type === "heading") return `${"#".repeat(node.level)} ${inlineToMarkdown(node.text)}`;
    if (node.type !== "block") return "";

    if (node.name === "summary" || node.name === "paragraph") return inlineToMarkdown(node.text);
    if (node.name === "skill") return "";
    if (node.name === "constraint") return labelledQuoteToMarkdown("Constraint", node.text);
    if (node.name === "decision") return `### ${node.attrs.id}\n\n**Decision:** ${inlineToMarkdown(node.text)}`;
    if (node.name === "task") return `- [${node.attrs.status === "done" ? "x" : " "}] ${inlineToMarkdown(node.text)}`;
    if (node.name === "file") return `**File:** \`${node.attrs.path}\`\n\n${inlineToMarkdown(node.text)}`;
    if (node.name === "api") return `**API:** \`${node.attrs.name}\`\n\n${inlineToMarkdown(node.text)}`;
    if (node.name === "link") return `[${node.text || node.attrs.href}](${node.attrs.href})`;
    if (node.name === "metadata") return `<!-- lessmark:${node.attrs.key}=${node.text} -->`;
    if (node.name === "risk") return labelledQuoteToMarkdown(`Risk (${node.attrs.level})`, node.text);
    if (node.name === "depends-on") return labelledQuoteToMarkdown(`Depends on \`${node.attrs.target}\``, node.text);
    if (node.name === "code") return `\`\`\`${node.attrs.lang ?? ""}\n${node.text}\n\`\`\``;
    if (node.name === "math") return mathToMarkdown(node.attrs.notation, node.text);
    if (node.name === "diagram") return `\`\`\`${node.attrs.kind}\n${node.text}\n\`\`\``;
    if (node.name === "example") return `Example:\n\n${node.text}`;
    if (node.name === "separator") return "---";
    if (node.name === "page" || node.name === "toc") return "";
    if (node.name === "nav") {
      if (!isSafeHref(node.attrs.href)) throw new Error("@nav href must be http, https, mailto, or a safe relative path");
      return `- [${inlineToMarkdown(node.attrs.label)}](${node.attrs.href})`;
    }
    if (node.name === "quote") return quoteToMarkdown(node.text, node.attrs.cite);
    if (node.name === "callout") return calloutToMarkdown(node.attrs.kind, node.attrs.title, node.text);
    if (node.name === "list") return listToMarkdown(node.attrs.kind, node.text);
    if (node.name === "table") return tableToMarkdown(node.attrs.columns, node.text);
    if (node.name === "image") return imageToMarkdown(node.attrs, node.text);
    if (node.name === "footnote") return `[^${node.attrs.id}]: ${inlineToMarkdown(node.text)}`;
    if (node.name === "definition") return `**${inlineToMarkdown(node.attrs.term)}**\n: ${inlineToMarkdown(node.text)}`;
    if (node.name === "reference") {
      const anchor = footnoteIds.has(node.attrs.target) ? `fn-${node.attrs.target}` : node.attrs.target;
      return `[${inlineToMarkdown(node.attrs.label || node.text || node.attrs.target)}](#${anchor})`;
    }
    return node.text;
  });
  return `${chunks.filter(Boolean).join("\n\n")}\n`;
}

function collectFootnoteIds(ast) {
  return new Set(ast.children
    .filter((node) => node.type === "block" && node.name === "footnote")
    .map((node) => node.attrs.id));
}

function fencedCodeNode(lang, body) {
  const text = body.join("\n");
  if (lang === "math" || lang === "tex" || lang === "latex") {
    return { type: "block", name: "math", attrs: { notation: "tex" }, text };
  }
  if (lang === "asciimath") {
    return { type: "block", name: "math", attrs: { notation: "asciimath" }, text };
  }
  if (["mermaid", "graphviz", "plantuml"].includes(lang)) {
    return { type: "block", name: "diagram", attrs: { kind: lang }, text };
  }
  return { type: "block", name: "code", attrs: lang ? { lang } : {}, text };
}

function mathToMarkdown(notation, text) {
  if (notation === "tex") return `$$\n${text}\n$$`;
  return `\`\`\`${notation}\n${text}\n\`\`\``;
}

function inlineToMarkdown(text) {
  let result = String(text);
  assertInlineLocalTargets(result);
  const replacements = [
    [/\{\{strong:([^{}]+)\}\}/g, "**$1**"],
    [/\{\{em:([^{}]+)\}\}/g, "*$1*"],
    [/\{\{code:([^{}]+)\}\}/g, "`$1`"],
    [/\{\{kbd:([^{}]+)\}\}/g, "`$1`"],
    [/\{\{del:([^{}]+)\}\}/g, "~~$1~~"],
    [/\{\{mark:([^{}]+)\}\}/g, "==$1=="],
    [/\{\{sup:([^{}]+)\}\}/g, "^$1^"],
    [/\{\{sub:([^{}]+)\}\}/g, "~$1~"],
    [/\{\{ref:([^{}|]+)\|([^{}]+)\}\}/g, "[$1](#$2)"],
    [/\{\{footnote:([^{}]+)\}\}/g, "[^$1]"],
    [/\{\{link:([^{}|]+)\|([^{}]+)\}\}/g, "[$1]($2)"]
  ];
  for (let pass = 0; pass < MAX_INLINE_MARKDOWN_PASSES; pass += 1) {
    const before = result;
    for (const [pattern, replacement] of replacements) {
      result = result.replace(pattern, replacement);
    }
    if (result === before) return result;
  }
  throw new Error("Inline nesting too deep");
}

function assertInlineLocalTargets(text) {
  for (const match of text.matchAll(/\{\{ref:[^{}|]*\|([^{}]*)\}\}/g)) {
    if (!DECISION_ID_PATTERN.test(match[1])) {
      throw new Error("Inline ref target must be a lowercase slug");
    }
  }
  for (const match of text.matchAll(/\{\{footnote:([^{}]*)\}\}/g)) {
    if (!DECISION_ID_PATTERN.test(match[1])) {
      throw new Error("Inline footnote target must be a lowercase slug");
    }
  }
}

function quoteToMarkdown(text, cite) {
  const quoted = inlineToMarkdown(text)
    .split("\n")
    .map((line) => `> ${line}`)
    .join("\n");
  return cite ? `${quoted}\n>\n> Source: ${inlineToMarkdown(cite)}` : quoted;
}

function labelledQuoteToMarkdown(label, text) {
  const lines = inlineToMarkdown(text).split("\n");
  const [first = "", ...rest] = lines;
  return [`> ${label}: ${first}`, ...rest.map((line) => `> ${line}`)].join("\n");
}

function calloutToMarkdown(kind, title, text) {
  const label = String(kind || "note").toUpperCase();
  const head = title ? `> [!${label}] ${inlineToMarkdown(title)}` : `> [!${label}]`;
  const body = inlineToMarkdown(text)
    .split("\n")
    .map((line) => `> ${line}`)
    .join("\n");
  return `${head}\n${body}`;
}

function listToMarkdown(kind, text) {
  const counters = [];
  return String(text)
    .split("\n")
    .filter((line) => line.trim() !== "")
    .map((line) => {
      const match = /^( *)- (.*)$/.exec(line);
      if (!match) throw new Error("@list items must use one explicit '- ' item marker per line");
      const level = match[1].length / 2;
      counters[level] = (counters[level] || 0) + 1;
      counters.length = level + 1;
      const marker = kind === "ordered" ? `${counters[level]}.` : "-";
      const item = inlineToMarkdown(match[2].trim());
      return `${"  ".repeat(level)}${marker} ${item}`;
    })
    .join("\n");
}

function tableToMarkdown(columns, text) {
  const header = splitTableRow(columns).map(escapeMarkdownTableCell);
  const rows = String(text)
    .split("\n")
    .filter((line) => line.trim() !== "")
    .map((line) => splitTableRow(line).map((cell) => escapeMarkdownTableCell(inlineToMarkdown(cell))));
  const table = [`| ${header.join(" | ")} |`, `| ${header.map(() => "---").join(" | ")} |`];
  for (const row of rows) table.push(`| ${row.join(" | ")} |`);
  return table.join("\n");
}

function escapeMarkdownTableCell(cell) {
  return String(cell).replace(/\|/g, "\\|");
}

function imageToMarkdown(attrs, text) {
  const image = `![${attrs.alt}](${attrs.src})`;
  return attrs.caption || text ? `${image}\n\n${inlineToMarkdown(attrs.caption || text)}` : image;
}

function readImageLine(line) {
  const match = /^!\[([^\]]*)\]\((\S+?)(?:\s+"([^"]*)")?\)\s*$/.exec(line.trim());
  if (!match) return null;
  return { alt: match[1], src: match[2], caption: match[3] || "" };
}

function readBlockquote(lines, startIndex) {
  if (!/^\s*>\s?/.test(lines[startIndex])) return null;
  const quoteLines = [];
  let index = startIndex;
  while (index < lines.length && /^\s*>\s?/.test(lines[index])) {
    quoteLines.push(lines[index].replace(/^\s*>\s?/, ""));
    index += 1;
  }

  const first = quoteLines[0]?.trim() ?? "";
  const callout = /^\[!(NOTE|TIP|WARNING|CAUTION)\]\s*(.*)$/i.exec(first);
  if (callout) {
    const attrs = { kind: callout[1].toLowerCase() };
    if (callout[2].trim()) attrs.title = plainText(callout[2]);
    return {
      nextIndex: index,
      node: {
        type: "block",
        name: "callout",
        attrs,
        text: quoteLines.slice(1).map(plainText).filter(Boolean).join("\n")
      }
    };
  }

  return {
    nextIndex: index,
    node: {
      type: "block",
      name: "quote",
      attrs: {},
      text: quoteLines.map(plainText).filter(Boolean).join("\n")
    }
  };
}

function readMarkdownList(lines, startIndex) {
  const first = readMarkdownListItem(lines[startIndex]);
  if (!first) return null;
  const kind = first.kind;
  const marker = first.marker;
  const items = [];
  let index = startIndex;
  while (index < lines.length) {
    const item = readMarkdownListItem(lines[index]);
    if (!item || item.kind !== kind) break;
    if (item.marker !== marker) {
      throw new Error("Mixed Markdown list markers are not supported by Lessmark import");
    }
    items.push(`${"  ".repeat(item.level)}- ${plainText(item.text)}`);
    index += 1;
  }
  return {
    nextIndex: index,
    node: {
      type: "block",
      name: "list",
      attrs: { kind },
      text: items.join("\n")
    }
  };
}

function readMarkdownListItem(line) {
  const match = /^( *)(?:([-*+])|(\d+[.)]))\s+(.+?)\s*$/.exec(line);
  if (!match) return null;
  return {
    level: Math.floor(match[1].length / 2),
    kind: match[3] ? "ordered" : "unordered",
    marker: match[2] || match[3].replace(/\d+/g, "1"),
    text: match[4]
  };
}

function readTable(lines, startIndex) {
  if (startIndex + 1 >= lines.length) return null;
  const header = splitMarkdownTableRow(lines[startIndex]);
  const separators = splitMarkdownTableRow(lines[startIndex + 1]);
  if (header.length < 1 || header.length !== separators.length || !separators.every(isTableSeparator)) return null;

  const rows = [];
  let index = startIndex + 2;
  while (index < lines.length && lines[index].trim() !== "") {
    const cells = splitMarkdownTableRow(lines[index]);
    if (cells.length !== header.length) break;
    rows.push(cells);
    index += 1;
  }

  const columns = header.map((cell) => escapeLessmarkTableCell(plainText(cell)));
  const body = rows.map((row) => row.map((cell) => escapeLessmarkTableCell(plainText(cell))));
  if (columns.some((cell) => cell === "") || body.some((row) => row.some((cell) => cell === ""))) return null;

  return {
    nextIndex: index,
    node: {
      type: "block",
      name: "table",
      attrs: { columns: columns.join("|") },
      text: body.map((row) => row.join("|")).join("\n")
    }
  };
}

function splitMarkdownTableRow(line) {
  let row = line.trim();
  if (row.startsWith("|")) row = row.slice(1);
  if (row.endsWith("|") && !row.endsWith("\\|")) row = row.slice(0, -1);
  return splitTableRow(row);
}

function isTableSeparator(cell) {
  return /^:?-{3,}:?$/.test(cell.trim());
}

function escapeLessmarkTableCell(cell) {
  return String(cell).replace(/\\/g, "\\\\").replace(/\|/g, "\\|");
}

function isMarkdownBlockStart(lines, index) {
  return /^(#{1,6})\s+/.test(lines[index]) ||
    readFenceLine(lines[index]) !== null ||
    readMathFenceLine(lines[index]) ||
    isMarkdownSeparator(lines[index]) ||
    /^\s*[-*]\s+\[[ xX]\]\s+/.test(lines[index]) ||
    readMarkdownListItem(lines[index]) !== null ||
    readImageLine(lines[index]) !== null ||
    /^\s*>\s?/.test(lines[index]) ||
    readTable(lines, index) !== null;
}

function isMarkdownSeparator(line) {
  return /^(?: {0,3})([-*_])(?:\s*\1){2,}\s*$/.test(line);
}

function readFenceLine(line) {
  const match = /^( {0,3})(`{3,}|~{3,})(.*)$/.exec(line);
  if (!match) return null;
  const marker = match[2];
  const info = match[3].trim();
  if (marker[0] === "`" && info.includes("`")) return null;
  return {
    char: marker[0],
    length: marker.length,
    lang: /^[A-Za-z0-9_.+-]+$/.test(info.split(/\s+/, 1)[0] ?? "") ? info.split(/\s+/, 1)[0] : ""
  };
}

function readMathFenceLine(line) {
  return line.trim() === "$$";
}

function isClosingFence(line, fence) {
  const match = /^( {0,3})(`{3,}|~{3,})\s*$/.exec(line);
  return Boolean(match && match[2][0] === fence.char && match[2].length >= fence.length);
}

function plainText(text) {
  return String(text)
    .replace(/!\[([^\]]*)\]\([^)]+\)/g, "$1")
    .replace(/\[([^\]]+)\]\([^)]+\)/g, "$1")
    .replace(/[`*_~]/g, "")
    .trim();
}

File: crates/lessmark/src/markdown.rs
use crate::ast::{Document, Node};
use crate::error::LessmarkError;
use crate::format::format_document;
use crate::parser::parse_lessmark;
use regex::Regex;
use std::collections::BTreeMap;

struct Fence {
    marker: char,
    length: usize,
    lang: String,
}

pub fn from_markdown(markdown: &str) -> Result<String, LessmarkError> {
    let normalized = markdown
        .trim_start_matches('\u{feff}')
        .replace("\r\n", "\n")
        .replace('\r', "\n");
    let lines = normalized.split('\n').collect::<Vec<_>>();
    let mut children = Vec::new();
    let mut index = 0;
    let mut first_paragraph = true;

    while index < lines.len() {
        let line = lines[index];

        if line.trim().is_empty() {
            index += 1;
            continue;
        }

        if let Some((level, text)) = read_heading(line) {
            children.push(Node::Heading {
                level,
                text: plain_text(text),
                position: None,
            });
            index += 1;
            continue;
        }

        if let Some(fence) = read_fence_line(line) {
            let mut body = Vec::new();
            index += 1;
            while index < lines.len() && !is_closing_fence(lines[index], &fence) {
                body.push(escape_block_line(lines[index]));
                index += 1;
            }
            if index >= lines.len() {
                return Err(LessmarkError::new(
                    "Unclosed fenced code block",
                    lines.len(),
                    1,
                ));
            }
            index += 1;

            let mut attrs = BTreeMap::new();
            if !fence.lang.is_empty() {
                attrs.insert("lang".to_string(), fence.lang);
            }
            children.push(Node::Block {
                name: "code".to_string(),
                attrs,
                text: body.join("\n"),
                position: None,
            });
            first_paragraph = false;
            continue;
        }

        if let Some((done, text)) = read_task(line) {
            let mut attrs = BTreeMap::new();
            attrs.insert(
                "status".to_string(),
                if done { "done" } else { "todo" }.to_string(),
            );
            children.push(Node::Block {
                name: "task".to_string(),
                attrs,
                text: plain_text(text),
                position: None,
            });
            first_paragraph = false;
            index += 1;
            continue;
        }

        let mut paragraph = Vec::new();
        while index < lines.len() && !lines[index].trim().is_empty() {
            if read_heading(lines[index]).is_some()
                || read_fence_line(lines[index]).is_some()
                || read_task(lines[index]).is_some()
            {
                break;
            }
            paragraph.push(lines[index].trim());
            index += 1;
        }

        let raw_text = paragraph.join(" ");
        if let Some((label, href)) = read_standalone_link(&raw_text) {
            let mut attrs = BTreeMap::new();
            attrs.insert("href".to_string(), href.to_string());
            children.push(Node::Block {
                name: "link".to_string(),
                attrs,
                text: label.to_string(),
                position: None,
            });
        } else {
            children.push(Node::Block {
                name: if first_paragraph { "summary" } else { "note" }.to_string(),
                attrs: BTreeMap::new(),
                text: plain_text(&raw_text),
                position: None,
            });
        }
        first_paragraph = false;
    }

    format_document(&Document::new(children)).map_err(|errors| {
        LessmarkError::new(
            errors
                .first()
                .map(|error| error.message.clone())
                .unwrap_or_else(|| "Cannot format imported Markdown".to_string()),
            1,
            1,
        )
    })
}

pub fn to_markdown(lessmark: &str) -> Result<String, LessmarkError> {
    let ast = parse_lessmark(lessmark)?;
    let chunks = ast
        .children
        .iter()
        .filter_map(markdown_node)
        .filter(|chunk| !chunk.is_empty())
        .collect::<Vec<_>>();
    Ok(format!("{}\n", chunks.join("\n\n")))
}

fn markdown_node(node: &Node) -> Option<String> {
    match node {
        Node::Heading { level, text, .. } => {
            Some(format!("{} {}", "#".repeat(*level as usize), text))
        }
        Node::Block {
            name, attrs, text, ..
        } => match name.as_str() {
            "summary" | "note" => Some(text.clone()),
            "warning" => Some(format!("> Warning: {}", text)),
            "constraint" => Some(format!("> Constraint: {}", text)),
            "decision" => Some(format!(
                "**Decision {}:** {}",
                attrs.get("id").map(String::as_str).unwrap_or(""),
                text
            )),
            "task" => Some(format!(
                "- [{}] {}",
                if attrs.get("status").map(String::as_str) == Some("done") {
                    "x"
                } else {
                    " "
                },
                text
            )),
            "file" => Some(format!(
                "**File:** `{}`\n\n{}",
                attrs.get("path").map(String::as_str).unwrap_or(""),
                text
            )),
            "api" => Some(format!(
                "**API:** `{}`\n\n{}",
                attrs.get("name").map(String::as_str).unwrap_or(""),
                text
            )),
            "link" => Some(format!(
                "[{}]({})",
                if text.is_empty() {
                    attrs.get("href").map(String::as_str).unwrap_or("")
                } else {
                    text
                },
                attrs.get("href").map(String::as_str).unwrap_or("")
            )),
            "metadata" => Some(format!(
                "<!-- lessmark:{}={} -->",
                attrs.get("key").map(String::as_str).unwrap_or(""),
                text
            )),
            "risk" => Some(format!(
                "> Risk ({}): {}",
                attrs.get("level").map(String::as_str).unwrap_or(""),
                text
            )),
            "depends-on" => Some(format!(
                "> Depends on `{}`: {}",
                attrs.get("target").map(String::as_str).unwrap_or(""),
                text
            )),
            "code" => Some(format!(
                "```{}\n{}\n```",
                attrs.get("lang").map(String::as_str).unwrap_or(""),
                text
            )),
            "example" => Some(format!("Example:\n\n{}", text)),
            _ => Some(text.clone()),
        },
    }
}

fn read_heading(line: &str) -> Option<(u8, &str)> {
    let marker_count = line.chars().take_while(|char| *char == '#').count();
    if !(1..=6).contains(&marker_count) || line.chars().nth(marker_count) != Some(' ') {
        return None;
    }
    let text = line[marker_count + 1..].trim();
    if text.is_empty() {
        return None;
    }
    Some((marker_count as u8, text))
}

fn read_task(line: &str) -> Option<(bool, &str)> {
    let re = Regex::new(r"^\s*[-*]\s+\[([ xX])\]\s+(.+?)\s*$").expect("task regex compiles");
    let captures = re.captures(line)?;
    Some((
        captures.get(1)?.as_str().eq_ignore_ascii_case("x"),
        captures.get(2)?.as_str(),
    ))
}

fn read_standalone_link(text: &str) -> Option<(&str, &str)> {
    let re = Regex::new(r"^\[([^\]]+)\]\((https?://[^)\s]+|mailto:[^)\s]+)\)$")
        .expect("link regex compiles");
    let captures = re.captures(text)?;
    Some((captures.get(1)?.as_str(), captures.get(2)?.as_str()))
}

fn read_fence_line(line: &str) -> Option<Fence> {
    let leading_spaces = line.chars().take_while(|char| *char == ' ').count();
    if leading_spaces > 3 {
        return None;
    }
    let rest = &line[leading_spaces..];
    let marker = rest.chars().next()?;
    if marker != '`' && marker != '~' {
        return None;
    }
    let length = rest.chars().take_while(|char| *char == marker).count();
    if length < 3 {
        return None;
    }
    let info = rest[length..].trim();
    if marker == '`' && info.contains('`') {
        return None;
    }
    let first_word = info.split_whitespace().next().unwrap_or("");
    let lang_re = Regex::new(r"^[A-Za-z0-9_.+-]+$").expect("language regex compiles");
    Some(Fence {
        marker,
        length,
        lang: if lang_re.is_match(first_word) {
            first_word.to_string()
        } else {
            String::new()
        },
    })
}

fn is_closing_fence(line: &str, fence: &Fence) -> bool {
    let leading_spaces = line.chars().take_while(|char| *char == ' ').count();
    if leading_spaces > 3 {
        return false;
    }
    let rest = &line[leading_spaces..];
    let length = rest
        .chars()
        .take_while(|char| *char == fence.marker)
        .count();
    length >= fence.length && rest[length..].trim().is_empty()
}

fn escape_block_line(line: &str) -> String {
    if line.starts_with('#') || line.starts_with('@') {
        format!("  {}", line)
    } else {
        line.to_string()
    }
}

fn plain_text(text: &str) -> String {
    let image_re = Regex::new(r"!\[([^\]]*)\]\([^)]+\)").expect("image regex compiles");
    let link_re = Regex::new(r"\[([^\]]+)\]\([^)]+\)").expect("link regex compiles");
    let markup_re = Regex::new(r"[`*_~]").expect("markup regex compiles");
    let without_images = image_re.replace_all(text, "$1");
    let without_links = link_re.replace_all(&without_images, "$1");
    markup_re.replace_all(&without_links, "").trim().to_string()
}
